lib/logmatcher.c

```diff
@@ -445,7 +447,17 @@ log_matcher_pcre_re_feed_named_substrings(LogMatcher *s, LogMessage *msg, int *m
           if (begin_index < 0 || end_index < 0)
             continue;
 
-          log_msg_set_value_by_name(msg, tabptr + 2, value + begin_index, end_index - begin_index);
+          if(self->super.prefix != NULL)
+            {
+              GString *formatted_key = g_string_new(self->super.prefix);
+              g_string_append(formatted_key, tabptr + 2);
+              log_msg_set_value_by_name(msg, formatted_key->str, value + begin_index, end_index - begin_index);
+              g_string_free(formatted_key, TRUE);
+            }
+          else
+            {
+              log_msg_set_value_by_name(msg, tabptr + 2, value + begin_index, end_index - begin_index);
+            }
```

Comment:
This can be a pretty hot path so we need to be aware of (a) performance implications. Also, this code is becoming a bit too complex, so (b) extracting the functionality into a separate function seems to be a good move.

Hints:
(a) g_string_new() is pretty slow, it allocates both a GString header and then the payload, which is potentially reallocated right at the next line. So we have 2-3 memory allocations plus a free operation in a path that is executed multiple times for each message. Message rates can be pretty high with syslog-ng, reaching 100s of thousands per second, although regexps would slow those down to a 10s of thousands. In any case, this is potentially executing 100k per sec. There's an API on top of GStrings that will give you cheap/free allocation of GString instances, look at ```scratch-buffers.h``` and try to find a user of that code for examples.

(b) you could move this part in its entirety into a local/static function, which would make this function a lot easier to read.